DEC/src/theories/ThDEC/semantics.jl

```diff
@@ -0,0 +1,73 @@
+
+"""
+function create_dynamic_model(sd, hodge)::Dict{TypedApplication, Any}
+  Dict{TypedApplication, Any}(
+    TA(*, Sort[Scalar(), Scalar()]) => 1,
+    TA(*, Sort[Scalar(), PrimalForm(0)]) => 1,
+    # Regular Hodge Stars
+    TA(★, Sort[PrimalForm(0)]) => Decapodes.dec_mat_hodge(0, sd, hodge),
+    TA(★, Sort[PrimalForm(1)]) => Decapodes.dec_mat_hodge(1, sd, hodge),
+    TA(★, Sort[PrimalForm(2)]) => Decapodes.dec_mat_hodge(2, sd, hodge),
+
+    # Inverse Hodge Stars
+    TA(★, Sort[DualForm(0)]) => Decapodes.dec_mat_inverse_hodge(0, sd, hodge), 
+    # TODO verify ^ why is this 1???
+    TA(★, Sort[DualForm(1)]) => Decapodes.dec_pair_inv_hodge(Val{1}, sd, hodge), 
+    # Special since Geo is a solver
+    TA(★, Sort[DualForm(2)]) => Decapodes.dec_mat_inverse_hodge(0, sd, hodge),
+
+    # Differentials
+    TA(d, Sort[PrimalForm(0)]) => Decapodes.dec_mat_differential(0, sd),
+    TA(d, Sort[PrimalForm(1)]) => Decapodes.dec_mat_differential(1, sd),
+
+    # Dual Differentials
+    TA(d, Sort[DualForm(0)]) => Decapodes.dec_mat_dual_differential(0, sd),
+    TA(d, Sort[DualForm(1)]) => Decapodes.dec_mat_dual_differential(1, sd),
+
+    # Wedge Products
+    TA(∧, Sort[PrimalForm(0), PrimalForm(1)]) => Decapodes.dec_pair_wedge_product(Tuple{0,1}, sd),
+    TA(∧, Sort[PrimalForm(1), PrimalForm(0)]) => Decapodes.dec_pair_wedge_product(Tuple{1,0}, sd),
+    TA(∧, Sort[PrimalForm(0), PrimalForm(2)]) => Decapodes.dec_pair_wedge_product(Tuple{0,2}, sd),
+    TA(∧, Sort[PrimalForm(2), PrimalForm(0)]) => Decapodes.dec_pair_wedge_product(Tuple{2,0}, sd),
+    TA(∧, Sort[PrimalForm(1), PrimalForm(1)]) => Decapodes.dec_pair_wedge_product(Tuple{1,1}, sd),
+
+    # Primal-Dual Wedge Products
+    TA(∧, Sort[PrimalForm(1), DualForm(1)]) => Decapodes.dec_wedge_product_pd(Tuple{1,1}, sd),
+    TA(∧, Sort[PrimalForm(0), DualForm(1)]) => Decapodes.dec_wedge_product_pd(Tuple{0,1}, sd),
+    TA(∧, Sort[PrimalForm(1), DualForm(1)]) => Decapodes.dec_wedge_product_dp(Tuple{1,1}, sd),
+    TA(∧, Sort[PrimalForm(1), DualForm(0)]) => Decapodes.dec_wedge_product_dp(Tuple{1,0}, sd),
+
+    # Dual-Dual Wedge Products
+    # TA(∧, Sort[DualForm(1), DualForm(1)]) => Decapodes.dec_wedge_product_dd(Tuple{1,1}, sd),
+    TA(∧, Sort[DualForm(1), DualForm(0)]) => Decapodes.dec_wedge_product_dd(Tuple{1,0}, sd),
+    TA(∧, Sort[DualForm(0), DualForm(1)]) => Decapodes.dec_wedge_product_dd(Tuple{0,1}, sd),
+
+    # # Dual-Dual Interior Products
+    TA(ι, Sort[DualForm(1), DualForm(1)]) => Decapodes.interior_product_dd(Tuple{1,1}, sd),
+    TA(ι, Sort[DualForm(1), DualForm(2)]) => Decapodes.interior_product_dd(Tuple{1,2}, sd),
+
+    # # Dual-Dual Lie Derivatives
+    # :ℒ₁ => ℒ_dd(Tuple{1,1}, sd)
+
+    # # Dual Laplacians
+    # :Δᵈ₀ => Δᵈ(Val{0},sd)
+    # :Δᵈ₁ => Δᵈ(Val{1},sd)
+
+    # # Musical Isomorphisms
+    TA(♯, Sort[PrimalForm(1)]) => Decapodes.dec_♯_p(sd), # Primal(1) -> PVField
```

Comment:
As discussed on Zulip and offline, this dict needs to be refactored to emit functions based on input and output types both. This is to accommodate e.g. the fact that there is a $$\flat_{pp}$$ and a $$\flat_{pd}$$. See [Figure 5.2](https://thesis.library.caltech.edu/1885/3/thesis_hirani.pdf).

Reply:
At this level, the output type is determined by the input types, so I'm not really sure what this means.

Reply:
In the DEC, an e.g. sharp operator can be defined which takes a primal form and outputs a primal vector field, while another sharp operator takes a primal form and emits a dual vector field.